### ISSUE TITLE
sample_tasks: tasks_seed independent of input_shard implicitly inflates context-per-task duplication under multi-shard sweeps

### ISSUE THREAD

**Issue by mmcdermott** [payalchandak/EveryQuery#182]
## Summary

`sample_tasks.py` derives the task draw seed from `(seed, "tasks", task_shard)` and the context draw seed from `(seed, "contexts", input_shard, task_shard)` ([`sample_tasks.py#L518-L537`](https://github.com/payalchandak/EveryQuery/blob/main/src/every_query/generate_tasks/sample_tasks.py#L518-L537)). The script's docstring documents this as a deliberate symmetry: "Fixing `task_shard` and varying `input_shard` evaluates the *same* tasks on *different* patients."

That symmetry, combined with the typical sweep shape from [`configs/hydra/launcher/cpu_array.yaml`](https://github.com/payalchandak/EveryQuery/blob/main/configs/hydra/launcher/cpu_array.yaml) (`input_shard=0..K−1 × task_shard=0..T−1`), means a `K × T` worker grid produces only `T × n_tasks` **unique** `(code, duration_days)` pairs even though it writes `K × T × n_tasks × contexts_per_task` rows. Each unique task is duplicated `K × contexts_per_task` times along the patient axis. For an arbitrary-query pretraining objective like EQ's, that is an under-investment in task variety: the model's query-embedding coverage is bounded by `T × n_tasks`, regardless of how many input shards the sweep walks.

This is a **design-shape issue, not a determinism bug**. Reproduction is two function calls:

```python
from every_query.utils.seeds import derive_seed
from every_query.generate_tasks.sample_tasks import sample_tasks
codes = [f"CODE_{i:04d}" for i in range(1000)]
a = sample_tasks(1024, codes, 1, 731, seed=derive_seed(1, "tasks", 0))   # input_shard=0, task_shard=0
b = sample_tasks(1024, codes, 1, 731, seed=derive_seed(1, "tasks", 0))   # input_shard=7, task_shard=0
assert a == b   # passes — input_shard never reached the seed
```

Simulating a `K=8 × T=16` sweep with `n_tasks=1024, contexts_per_task=1`:

```
Workers:                  128
Total task draws:         131,072
Unique (code, duration):   14,414  ← 8.7× duplication on the task axis
Effective M (unique):      14,414
Effective N (per task):    8       (one context per input_shard)
```

The same compute budget under a non-redundant seeding scheme would give M = 131,072, N = 1.

## Why this matters for EQ pretraining

EQ's loss conditions on `(patient_history, query_code, duration)`. The query embedding is shared with the input vocab, so each query code already gets thousands of gradient signals from being a *context* token. The marginal value of a second exposure of the same `(code, duration)` query token paired with a new patient is small relative to the marginal value of a brand-new `(code, duration)` query — particularly for under-sampled regions of the query × duration space.

There is a calibration argument for `N > 1`: rare-event tasks (prevalence < ~0.1%) need multiple exposures so the model learns the right base rate. But that's an argument for *prevalence-stratified oversampling on the rare tail*, not blanket `N > 1` across the whole task universe.

The current code's M/N tradeoff is implicit — it falls out of `(K, T, n_tasks, contexts_per_task)` interactions rather than being directly configurable. That makes it easy to misconfigure without realizing it (a sweep that looks like "more compute" on the surface may actually be amplifying duplication).

## How this relates to existing issues

- **#149 (determinism audit)** lists the seed-axis split as an *invariant*. This issue argues the *split itself* is doing semantic work that shouldn't be implicit. Fixing it changes the determinism contract slightly (specifically, it removes the same-task-different-patient guarantee), so the two issues need to be coordinated.
- **#144 (TaskDistribution unification)** and **#148 (sampling pipeline)** are the natural home for the holistic redesign. A `random` stage with an explicit `n` (total realized tasks) and no per-shard worker structure makes the M/N choice a single config parameter rather than an emergent property of the sweep shape.
- **#125 (hypothesis tests)** could grow a property test that asserts the realized-task-set cardinality matches the configured budget — would have caught this design quirk regardless of which fix is applied.

### Proposal 1 — Minimal patch (1 line, ~15 min)

Add `input_shard` to the tasks_seed:

```diff
-    tasks_seed = derive_seed(seed, "tasks", task_shard)
+    tasks_seed = derive_seed(seed, "tasks", input_shard, task_shard)
```

Effect: every `(input_shard, task_shard)` worker draws independently, so a `K × T` sweep produces `K × T × n_tasks` unique tasks. No interface change. The "same tasks on different patients" symmetry that the docstring describes goes away — but for the pretraining objective it isn't load-bearing, and removing it is exactly the goal here.

Tests to add:
- `tests/test_sample_tasks.py::test_input_shard_changes_task_draw` — assert `sample_tasks` outputs differ across `input_shard` for the same `task_shard` and `seed`.
- Update any existing test that asserts the cross-shard symmetry to assert the new contract instead.

Tradeoffs:
- (+) Trivially small, lands today, immediately recovers `K×` task variety on every existing sweep without any other configuration change.
- (−) Doesn't expose M/N as a first-class config knob. A user who *wants* the same task seen on N patient contexts (e.g. for the rare-event calibration argument) loses an easy way to do that.
- (−) Changes the seed contract — anyone with task parquets generated by the current code who reruns will get different task sets even with the same `seed`. Worth a brief note in CHANGELOG.

### Proposal 2 — Explicit M/N controls in `sample_tasks_config.yaml` (medium, ~half a day)

Make the budget shape explicit in the per-worker config:

```yaml
# Per-worker sampling budget (interpretation now explicit)
budget:
  unique_tasks: 1024              # M — number of distinct (code, duration) pairs to sample
  contexts_per_task: 1            # N — number of patient contexts per task
  cross_shard_task_diversity: per_shard   # one of: per_shard | global
```

Where `cross_shard_task_diversity`:

- `per_shard` (Proposal 1's behaviour): `tasks_seed` includes `input_shard`. Each worker draws unique tasks from scratch. The default for "I want maximum query variety."
- `global`: today's behaviour. Same `task_shard` index across workers draws the same tasks. Useful when you genuinely want "same M tasks evaluated on `K × N` distinct patient contexts" — e.g., a calibration ablation where you want the same query universe evaluated on every input shard.

This makes the M/N tradeoff a one-knob choice rather than a property of how the user invokes the Hydra sweep. Pairs cleanly with `cpu_array.yaml`'s sweep semantics — it documents what the sweep is actually producing.

Tests to add:
- `test_per_shard_diversity_max_unique_tasks` — assert unique-task count = `K × T × n_tasks`.
- `test_global_diversity_replicates_today` — assert unique-task count = `T × n_tasks` (matches today's behaviour).
- A `pytest.mark.parametrize` matrix over `(cross_shard_task_diversity, K, T, n_tasks, contexts_per_task)` to lock the cardinality contract for both modes.

Tradeoffs:
- (+) Surfaces the M/N decision in config; reviewer/maintainer can see at a glance what a config will produce.
- (+) Backwards-compat path via `global` mode for anyone who relied on the old behaviour.
- (−) Twice the config surface in `sample_tasks_config.yaml`. Mitigated by good defaults.

### Proposal 3 — Holistic: align with #148's sampling pipeline (the right long-term home)

Under #148, the per-worker `(input_shard, task_shard, n_tasks, contexts_per_task)` tuple goes away in favour of a top-level `sampling:` pipeline that names what the user actually wants:

```yaml
sampling:
  - mode: random
    n: 10_000_000               # total unique realized tasks across all shards
    distribution: base
    seed_split: train_random

  # Optional: prevalence-stratified oversampling for the rare-event tail
  - mode: random
    n: 500_000
    distribution: rare_only      # filter on prevalence column from #146 metadata file
    contexts_per_task: 4         # legitimate place for N>1 — calibration on rare cells
    seed_split: train_rare_calib
```

The pipeline runs once, the input_shard / task_shard partitioning becomes an implementation detail of how the workload is parallelized rather than a semantic axis the user has to reason about. The bug in this issue stops being possible because there's no longer an `(input_shard, task_shard)` × `n_tasks` shape to misconfigure.

This is the right shape for the codebase's direction (#144, #148) but it's a much bigger lift. It also makes it obvious where the rare-event N>1 calibration argument lives (a separate sampling stage), without polluting the global default.

I'd suggest landing **Proposal 1 now** as a stop-the-bleeding fix to recover task variety on existing sweeps, **Proposal 2 next** as the bridge that makes the M/N decision explicit, and **Proposal 3 as part of the #144/#148 work** that's already on the roadmap.

## Logging / observability gap (cross-cutting)

Independent of which proposal lands, this design issue would have been caught much earlier — and post-hoc analysis on existing runs would be much cheaper — if the task generator and trainer surfaced sampling cardinality as first-class telemetry. Concrete proposals:

### At task-generation time

- Each worker writes a small JSON sidecar next to its parquet:

  ```json
  {
    "input_shard": "0",
    "task_shard": 0,
    "seed": 1,
    "n_tasks_drawn": 1024,
    "contexts_per_task": 1,
    "unique_tasks": 1024,
    "n_rows_written": 1024,
    "n_subjects": 873,
    "n_unique_prediction_times": 1019,
    "tasks_seed_inputs": ["seed=1", "task_shard=0"],
    "contexts_seed_inputs": ["seed=1", "input_shard=0", "task_shard=0"]
  }
  ```

  At minimum the seed inputs — that single field would have fully resolved the ambiguity that motivated this issue.

- After the full sweep finishes, an aggregator dumps a summary at the `task_labels_dir` root:

  ```
  task_labels_dir/_sweep_summary.json
  ----------------------------------
  total_workers:      128
  total_rows:         131,072
  unique_tasks:       14,414       ← what the model actually has access to
  unique_subjects:    102,331
  unique_(s, pt):     287,419
  unique_codes:       8,217
  unique_durations:   621
  cross_shard_task_diversity_mode: global   # which seed contract was used
  ```

  This is what makes "did our sampling actually cover what we thought it would?" a one-grep question rather than a multi-day post-mortem.

### At training time

- `EveryQueryLightningModule.setup` reads `task_labels_dir/_sweep_summary.json` (if present) and logs it as wandb summary fields — `pretrain/unique_tasks`, `pretrain/unique_subjects`, `pretrain/cross_shard_task_diversity_mode`, etc. These get pinned to the run and are queryable from the wandb API for any post-hoc cohort comparison.
- Once a training step, log `train/sampled_query_count_in_batch` (number of distinct queries seen in the batch) and `train/sampled_duration_count_in_batch`. Cheap, and gives a per-step view of whether the dataloader is delivering the variety the sampler claimed to write.
- After training, persist a small `pretrain_sampling_summary.parquet` next to `best_model.ckpt` with the per-step distinct-query counts. Lets analyses like the one that motivated this issue read directly off the run dir without re-deriving anything from raw task parquets.

### Documented invariants

A short README under `src/every_query/generate_tasks/` that spells out:
- The seed axis split (which fields contribute to which seed)
- The cardinality formula (`unique_tasks = f(K, T, n_tasks, mode)`)
- A worked example for the recommended sweep
- A "common mistakes" appendix

## Asks

1. Open to a PR for **Proposal 1** into `dev`? Happy to send a minimal one (the seed change + tests + a short CHANGELOG note).
2. Is there appetite for **Proposal 2** as a follow-up before #144/#148 land, or do we want to wait and skip straight to #148's sampling pipeline?
3. Any objection to landing the **logging/observability** changes incrementally regardless of which proposal lands? Most of those are independent of the core seed-contract decision.

cc: this touches @mmcdermott's #149 invariants list directly — wanted to flag before sending the PR.
